presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorPageSink.java

```diff
@@ -208,34 +227,39 @@ public void appendPage(Page page)
         }
     }
 
-    private class BucketedPageWriter
+    private class PartitionedPageWriter
             implements PageWriter
     {
-        private final BucketFunction bucketFunction;
-        private final Int2ObjectMap<PageStore> pageStores = new Int2ObjectOpenHashMap<>();
+        private final Optional<BucketFunction> bucketFunction;
+        private final Optional<TemporalFunction> temporalFunction;
+        private final Long2ObjectMap<PageStore> pageStores = new Long2ObjectOpenHashMap<>();
 
-        public BucketedPageWriter(int bucketCount)
+        public PartitionedPageWriter()
         {
-            this.bucketFunction = new RaptorBucketFunction(bucketCount);
+            checkArgument(temporalColumnIndex.isPresent() == temporalColumnType.isPresent(),
+                    "temporalColumnIndex and temporalColumnType must be both present or absent");
+
+            this.bucketFunction = bucketCount.isPresent() ? Optional.of(new RaptorBucketFunction(bucketCount.getAsInt())) : Optional.empty();
+            this.temporalFunction = temporalColumnType.map(type -> TemporalFunction.create(temporalColumnType.get()));
         }
 
         @Override
         public void appendPage(Page page)
         {
-            Block[] blocks = new Block[bucketFields.length];
-            for (int i = 0; i < bucketFields.length; i++) {
-                blocks[i] = page.getBlock(bucketFields[i]);
-            }
-            Page bucketArgs = new Page(page.getPositionCount(), blocks);
+            Block temporalBlock = temporalColumnIndex.isPresent() ? page.getBlock(temporalColumnIndex.getAsInt()) : null;
```

Comment:
Instead of using a ternary, write it like this:
```java
Block temporalBlock = temporalColumnIndex.map(index -> page.getBlock(index)).orElse(null);